src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectFactory.cs

```diff
@@ -89,16 +101,19 @@ public VisualStudioProject CreateAndAddToWorkspace(string projectSystemName, str
                 if (w.CurrentSolution.ProjectIds.Count == 0)
                 {
                     // Fetch the current solution path. Since we're on the UI thread right now, we can do that.
-                    string? solutionFilePath = null;
-                    var solution = (IVsSolution)Shell.ServiceProvider.GlobalProvider.GetService(typeof(SVsSolution));
-                    if (solution != null)
+                    var solutionFilePath = _threadingContext.JoinableTaskFactory.Run(async () =>
                     {
-                        if (ErrorHandler.Failed(solution.GetSolutionInfo(out _, out solutionFilePath, out _)))
+                        await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+                        var solution = (IVsSolution)Shell.ServiceProvider.GlobalProvider.GetService(typeof(SVsSolution));
```

Comment:
@jasonmalinowski can we use IAsyncServiceProvider here?  I'm not really sure i understand where/when/how we can use async stuff from teh platform itself.

Reply:
Yep, that's fine. More or less wherever you see a regular service provider being used, IAsyncServiceProvider is no less safe. (I'm writing that carefully because we have places in Roslyn today where we use the regular service provider in unsafe ways, but even in those cases async provider isn't really less safe...)

Whether it's helpful is of course a different question. The IVsSolution service is defined by the shell and practically it would have already been loaded by somebody else first, so there wouldn't be a real perf win since it's just going to be synchronous.

Reply:
it's more about avoiding the exlicit switch to hte ui thread ourselves.